core/base/v7/inc/ROOT/RError.hxx

```diff
@@ -0,0 +1,230 @@
+   RError fError;
+public:
+   explicit RException(const RError &error) : std::runtime_error(error.GetReport()), fError(error) {}
+   const RError &GetError() const { return fError; }
+};
+
+
+/// Wrapper class that generates a data member of type T in RResult<T> for all Ts except T == void
+namespace Internal {
+template <typename T>
+struct RResultType {
+   RResultType() = default;
+   explicit RResultType(const T &value) : fValue(value) {}
+   T fValue;
+};
+
+class RResultTypeVoid {};
+} // namespace Internal
+
+
+// clang-format off
+/**
+\class ROOT::Experimental::RResult
+\ingroup Base
+\brief The class is used as a return type for operations that can fail; wraps a value of type T or an RError
+
+The RResult enforces checking whether it contains a valid value or an error state. If the RResult leaves the scope
+unchecked, it will throw an exception.  RResult should only be allocated on the stack, which is helped by deleting the
+new operator.  RResult is movable but not copyable to avoid throwing exceptions multiple times.
+*/
+// clang-format on
+template <typename T>
+class RResult : public std::conditional_t<std::is_void<T>::value, Internal::RResultTypeVoid, Internal::RResultType<T>> {
+private:
+   /// This is the nullptr for an RResult representing success
+   std::unique_ptr<RError> fError;
+   /// This is safe because checking an RResult is not a multi-threaded operation
+   mutable bool fIsChecked{false};
+
+public:
+   /// Constructor is _not_ explicit in order to allow for `return T();` for functions returning RResult<T>
+   /// Only available if T is not void
+   template <typename Dummy = T, typename = typename std::enable_if_t<!std::is_void<T>::value, Dummy>>
+   RResult(const Dummy &value) : Internal::RResultType<T>(value) {}
+   /// Constructor is _not_ explicit such that the RError returned by R__FAIL can be converted into an RResult<T>
+   /// for any T
+   RResult(const RError &error) : fError(std::make_unique<RError>(error)) {}
+
+   /// RResult<void> has a default contructor that creates an object representing success
+   template <typename Dummy = T, typename = typename std::enable_if_t<std::is_void<T>::value, Dummy>>
+   RResult() {}
+
+   RResult(const RResult &other) = delete;
+   RResult(RResult &&other) = default;
+   RResult &operator =(const RResult &other) = delete;
+   RResult &operator =(RResult &&other) = default;
+
+   ~RResult() noexcept(false)
+   {
+      if (R__unlikely(fError && !fIsChecked)) {
+         // Prevent from throwing if the object is deconstructed in the course of stack unwinding for another exception
+#if __cplusplus >= 201703L
+         if (std::uncaught_exceptions() == 0)
+#else
+         if (!std::uncaught_exception())
+#endif
+         {
+            throw RException(*fError);
+         }
+      }
+   }
+
+   /// Only available if T is not void
+   template <typename Dummy = T>
+   typename std::enable_if_t<!std::is_void<T>::value, const Dummy &>
```

Comment:
The `Get()` method needs to have access to `fError` and `fIsChecked`, so pretty much to all of the constituents of `RResult`. An approach that avoids `enable_if` could simply be implemented by a  `RResult<void>` specialization (without any base class detour), which can be done of course but it would have some code duplication. What do you think?

Reply:
Actually, I just tried that: `RResult<void>` being a template specialization of `RResult<T>`.  Both `RResult<T>` and `RResult<void>` inherit from `Internal::RResultBase`, which implements common handling of the error case.  It looks to me nicer and simpler now and solves a few other issues because `enable_if` isn't required anymore.

A notable difference is that the `RResult<T>` data members now start with `fError` and `fChecked`, only then comes `fValue`.  I was wondering if we need to do anything specific for padding, but I'd believe the compiler does sensible padding based on the type of `T`...?

Reply:
Indeed, there should be no problem here. Plus all the offsets are statically known, so also perf-wise it's fine.